### PR TITLE
introduce jsonrpsee client abstraction + kill HTTP support.

### PR DESCRIPTION
**Pull request by niklasad1**
Brings in the `abstract async jsonrpc client` from jsonrpsee and removes `HTTP support`.

The benefit is that is possible for users can plug in their own transport layer as long the `transport` implements the traits `TransportSenderT` and `TransportReceiverT`. For example in memory to talk with an embedded substrate node...

Follow-up introduce `proc macros` from jsonrpsee.

### CODE REVIEW

File: Cargo.toml
Comment (niklasad1) [inline Cargo.toml:2]: move

**Comment by jsdw** [comment]
If you resolve the conflict, you might find that `test-runtime` either builds OK now without getting stuck, or at least tells you why it's getting stuck (possibly you don't have `substrate` on your PATH locally?) :)

**Comment by niklasad1** [comment]
> If you resolve the conflict, you might find that test-runtime either builds OK now without getting stuck, or at least tells you why it's getting stuck (possibly you don't have substrate on your PATH locally?) :)

I have substrate in my path but it's some issue with connection to the substrate process in the integration tests but I tried `submit_and_watch example` and that worked.

However, thanks merged to master let's see that the CI tells us :)

**Comment by sander2** [comment]
> Would that solve your issues w.r.t. using an embedded node? Then we can decouple the all those heavy substrate deps to another repo.

Yea, I think this should work great, thanks!

**Comment by gregdhill** [comment]
Just following up on this, we're still maintaining a [custom fork](https://github.com/interlay/subxt/tree/client) with the embedded client but we'd very much like to use origin without modification. Let me or @sander2 know if we can help push this through the gate!

**Comment by niklasad1** [comment]
@gregdhill 

I haven't checked this for a while sorry but I ran into a race-condition (I got a reponse for a subscription before the subscription call were answered) but I can take a look again this week.

Otherwise, we would have wait until the jsonrpsee branch is merged to substrate then it should work.

**Comment by niklasad1** [comment]
@gregdhill @sander2 

I tested it locally with embedded client and JSON-RPC related stuff works as intended but I got some issues metadata issues in the embedded client (not sure exactly why) when submitting extrinsics but the RPC related code works as intended.

In addition it's quite hard/tricky to maintain the `embedded client` because `subxt` depends on `substrate deps v4.0.0` and some crates such as `sc-service` are not released on crates.io so in practice I think one have to patch all substrate deps to make it compile. 

For example `substrate master` won't work with `subxt` because it broke `PairSigner`, thus I used the tag when `v4.0.0` was released, so you may have to fork subxt or maintain a separate branch for it to work with new polkadot releases or other features that you need (but I guess you already doing that)

To conclude, we will not include the embedded client in this repo but because we are migrating to the `jsonrpsee core client` you should be able to maintain the `embedded client` yourselves and copy to a few jsonrpsee snippets from [here](https://github.com/paritytech/subxt/pull/341/commits/a45bf143bb334cbefe1d58107181c78b81dd88b9) to get it running

File: Cargo.toml
Comment (niklasad1) [inline Cargo.toml:9]:
```suggestion
    "test-runtime"
```

File: subxt/Cargo.toml
Comment (niklasad1) [inline subxt/Cargo.toml:8]:
```suggestion
readme = "../README.md"
```

File: subxt/Cargo.toml
Comment (niklasad1) [inline subxt/Cargo.toml:14]:
```suggestion
```

File: test-runtime/build.rs
Comment (niklasad1) [inline test-runtime/build.rs:76]:
```suggestion
            // It might take a while for the substrate node that spin up the RPC server.
```

File: subxt/src/rpc.rs
Comment (niklasad1) [inline subxt/src/rpc.rs:246]: NOTE; this panics if the serialization fails

Comment (jsdw) [reply subxt/src/rpc.rs:246]: I guess there should be no reason for these things to fail to serialize (afaik it's just maps where you have to be careful with json serializing?)

Comment (niklasad1) [reply subxt/src/rpc.rs:246]: indeed,

I'm working on a follow-up to replace this with proc macros to get rid of this anyway, so view this as temporary.

Comment (ascjones) [reply subxt/src/rpc.rs:246]: Yes these are infallible

File: subxt/src/rpc.rs
Comment (ascjones) [inline subxt/src/rpc.rs:547]: nit: I'd like this named even shorter, perhaps just `ws_client`. Matter of taste though.

**Review by ascjones — APPROVED** [review]
Awesome!

File: subxt/src/rpc.rs
Comment (niklasad1) [inline subxt/src/rpc.rs:547]:
```suggestion
pub async fn ws_client(url: &str) -> Result<RpcClient, RpcError> {
```

File: subxt/src/client.rs
Comment (niklasad1) [inline subxt/src/client.rs:89]:
```suggestion
            crate::rpc::ws_client(url).await?
```

**Pull request merged**
